### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by Bhopss152**
Joshua Hess
J0shh3ss

### CODE REVIEW

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Reflection.Metadata.Ecma335;
+
+namespace CanHazFunny;
+
+public class Jester
+{
+    public IJokeService JokeService { get; }
+    public IJokeOutput JokeOutput { get; }
+    public Jester(JokeService? _jokeService,IJokeOutput? _jokeOutput)
```
Comment (alowe4) [inline CanHazFunny/CanHazFunny/Jester.cs:10]: Jester needs to take both interfaces as dependencies, this should be IJokeService and not JokeService.
the rest looks great, I noticed no warnings, it compiles and passes all tests.

**Review by alowe4 — COMMENTED** [review]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
    The JokeService will need to have the interface applied to it. ✔
    Create an implementation of the output interface that writes out the joke to the console. ✔
    Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ❌
    The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
    Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔

    Be sure you enable:
        Nullability for all projects ✔
        Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
        Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
    Ensure there are no errors or warnings (including code analysis warnings) ✔
    All of the above should be unit tested ✔
    Choose simplicity over complexity ✔

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 2]
```diff
@@ -1,10 +1,60 @@
...
+            Mock<IJokeService> mock = new();
+            //Act
+            mock.Setup(jokeService => jokeService.GetJoke()).Returns("Test Joke");
+            //Assert
+            Assert.AreEqual<string>(temp, mock.Object.GetJoke());
+
+        }
+
+        [TestMethod]
+        public void TellJoke_Checks_For_ChuckNorris()
```
Comment (mmoran314) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:45]: This works in cases where the original joke being used in the method contains "Chuck Norris", in which the program would still work and the test will pass. However the test may also pass if the first joke being tested does not contain "Chuck Norris". I would consider rewriting this test with Moq to make sure that when the program does receive "Chuck Norris" it is properly filtered.

**Review by mmoran314 — COMMENTED** [review]
Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔

Looks good to me for the most part. Would redo the test for the check chuck norris to ensure that it still works when given a chuck norris joke. As it is now there is potential that the first string being tested will not contain Chuck Norris and the test will pass.
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: CanHazFunny/CanHazFunny/IJokeOutput.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿using System;
+namespace CanHazFunny;
+
+public interface IJokeOutput
+{
+    public void Write(string output);
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/IJokeOutput.cs:6]:
```suggestion
    void Write(string output);
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Reflection.Metadata.Ecma335;
+
+namespace CanHazFunny;
+
+public class Jester
+{
+    public IJokeService JokeService { get; }
+    public IJokeOutput JokeOutput { get; }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:9]: Do these need to be publically accessible? Maybe make them private.
```suggestion
    private IJokeService _JokeService;
    private IJokeOutput JokeOutput;
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    public Jester(IJokeService? _jokeService,IJokeOutput? _jokeOutput)
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:10]:
```suggestion
    public Jester(IJokeService jokeService, IJokeOutput jokeOutput)
```
If you are going to throw a null exception, don't mark the parameters as nullable. Marking them as nullable here says you think it's okay to pass null into your constructor.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+	{
+        JokeService = _jokeService ?? throw new ArgumentNullException(nameof(_jokeService));
+        JokeOutput = _jokeOutput ?? throw new ArgumentNullException(nameof(_jokeOutput));
+    }
+    public string? Joke { get; set; }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:15]: Keep this scoped to the method you're using it in.

**Pull request closed without merging**